Improve type safety and fix comparators

ash2k opened the pull request:
Hello.
I improved type safety and while refactoring noticed that some methods, that take `Comparator`, are incorrectly use `extends` instead of `super`. That shows that not only type safety helps to write correct tests but also to catch bugs in the assertion library itself.

What do you think?

ash2k committed:
Improve type safety and fix comparators

joel-costigliola commented:
Hi Mikhail,

I'm troubled here, your changes have broken some of the examples of fest assertions project https://github.com/joel-costigliola/fest-examples.

Let's see one example that doesn't compile anymore :

``` java
assertThat(frodo).usingComparator(raceNameComparator).isEqualTo(sam);
```

Here frodo's type is `TolkienCharacter` and raceNameComparator's is `Comparator<TolkienCharacter>` , thus I think a fest user should be able to write such an assertion.

I'm no expert in generics ... do you think we can have both safer type and still be able to write this kind of assertion ?
If we need to stick with <?> wildcard only to allow that then I think it would be the way to go.

My point here is : we must first offer flexible assertions for fest users and if possible safe ones if it doesn't break the previous statement.

Anyway, your thoughts are very welcome.

ash2k committed:
More type safety and fixes

ash2k commented:
Hi, Joel.

I fixed this issue and some others. Take a look.
I introduced `usingElementComparator()` to set a comparator for elements of arrays/iterables (`usingComparator()` sets comparator for whole objects, not for their elements). 

Sorry, i'm very short in time now and the patch is very rough. Needs more type cleanup and javadocs.

joel-costigliola commented:
I will try to review your commit the next week.
Thanks for your hard work.

joel-costigliola commented:
Hey Mikhail,

I'm reviewing your work, it's great, I'm improving javadoc and I think it will be ok.

Huge thanks !

ash2k commented:
Hello, Joel!

I'm glad you found it useful. I'm actively using the library in several of my projects and i'm taking notes of what functionality can be added. I'm going to send some pull requests or at least open issues (as soon as i have spare time).

Cheers.

joel-costigliola commented:
Merged and available in 2.0M4 freshly released.

Thanks again
